### ISSUE TITLE
Is Embedding layer mainly used for sequences only?

### ISSUE THREAD

**Issue by kaushalshetty** [fchollet/keras#7607]
Is Embedding layer mainly used for text_to_sequence(texts) only? Can it be used to embed a tfidf or a one hot encoded version of text?

**Comment by mahnerak** [comment]
I'm not sure what you mean by saying "tf-idf" encoded version of text, but one hot encoded version of text is not supported by embedding layer as an input.
According documentation (https://keras.io/layers/embeddings/), the Embedding layer gets a tensor with shape `(batch_size, sequence_length)` with dtype `int32` and returns a tensor with shape `(batch_size, sequence_length, output_dim)` with dtype `floatX`.
Consider changing your data format from "one-hot" to "array of indices" before feeding to model.

**Comment by kaushalshetty** [comment]
Hi mahnerak ,
Thanks for the answer . Just wanted one more clarification. When we give the input as an array of indices we have to make sure that they are of the same length and that is  why we pad_sequence which simply pads using 'pre' or 'post' truncations. Wont this step involve truncating important words(features) also? Is there a way to determine which words are not important and then truncate?

**Comment by mahnerak** [comment]
Actually you can make these sequences of same length without any truncations.
The function `pad_sequences` (https://keras.io/preprocessing/sequence/#pad_sequences) can get optional argument `maxlen`.
If you provide `maxlen`, all the sequences having length greater than `maxlen` are going to be truncated.
**If you don't provide `maxlen` it takes maximum of all lengths, and thus you don't lose any data.**

Don't forget to make your input layer/tensor with variable length (provide `None` as length).

**Comment by kaushalshetty** [comment]
Awesome. Got it. Please correct me whether my understanding of the embedding layer is correct or not.

1) Randomly initialize a embedding matrix(size: (vocabulary_size ,embedding_dimension))
2) Feed the network with input as a sequence of integers(all input having the same size). Is this the first layer of the network ? So the number of nodes in the first embedding layer is equal to the number of indexes in the sequence.
3) The network then takes each index from the sequence of integers and indexes that row of the embedding matrix. Does it retrieve that row of the embedding matrix using a one hot encoded version of the word multiplied by the embedding matrix or a simple dictionary indexing ?
4) once all the rows are indexed it is then flattened(to a 2D tensor) so as to feed it into the next Dense layer(If a dense layer is used). Will the number of nodes in the Dense layer be len(sequence_of_index)*embedding dimension
5) After running through the epochs gradient descent happens and the weights of the embedding matrix will change.


Also if I have a dataset which includes numerical features as well as categoriclal features. How will i use them along with the textual features and give input to the network? Can I make use of the embedding layer in this case?

**Comment by mahnerak** [comment]
1. True, Embedding starts with random vectors, with shape `(vocabulary_size , embedding_dimension)`.
2. Yes, if you are using functional API, then first layer (tensor) will be `Input(shape=(None,), dtype='int32')`. Embedding layer should output tensor with shape `(None, embedding_dimension)`. (times axes match with input).
3. It uses indexing, using `K.gather`. (look https://keras.io/backend/#gather)
4. No, it's not flattened. In case of if you want to apply Dense layer after Embedding layer on temporal data (which I don't recommend) you have to fix some length (say 200) for `maxlen`s in and `Input(shape=(200,), dtype='int32')`. In this case it can infer automatically that ouput shape of Embedding layer is (200, embedding_dimension) so Flatten() layer may be used to make using Dense layer possible. I recommend using LSTM or GRU layer instead of Dense (it can work with variable length data).
5. Yes, they behave as layer weights and during training embedding matrix change w.r.t. loss function.

If you have some numerical and/or categorical features, where time axis sizes match, you can just preprocess them using LSTM or GRU layer (try to use units=embedding_dimension) then concatenate with embedding layer output and then build recurrent neural network on top of it.

**Comment by kaushalshetty** [comment]
Great. Thanks a lot.
